### PR TITLE
feat(storybook): storybook config projectType application (#2382)

### LINKED ISSUE #2382: Storybook Install: "Could not read apps/ng-nest/tsconfig.lib.json"
# Prerequisites

Please answer the following questions for yourself before submitting an issue.
**YOU MAY DELETE THE PREREQUISITES SECTION.**

- [x ] I am running the latest version
- [ x] I checked the documentation (nx.dev) and found no answer
- [x ] I checked to make sure that this issue has not already been filed
- [x ] I'm reporting the issue to the correct repository (not related to React, Angular or any dependency)

## Expected Behavior
The expected behavior is to be able to install Storybook integration given the documentation instructions.

## Current Behavior
The current behavior when following the instructions for installing Storybook integration ends with an error "Could not read apps/ng-nest/tsconfig.lib.json" (which makes no sense, because the only `tsconfig.lib.json` in the project is in the `libs/api-interfaces/tsconfig.lib.json`, and I do not believe there would ever be a `apps/ng-nest/tsconfig.lib.json` file).

## Failure Information (for bugs)

Please help provide information about the failure if this is a bug. If it is not a bug, please remove the rest of this template.

### Steps to Reproduce

Please provide detailed steps for reproducing the issue.

1. `npx create-nx-workspace@latest nx-test-workspace`
Workspace: angular-nest
Application Name: ng-nest
Default style: SCSS (.scss)
2. `cd nx-test-workspace`
3. `npx -p @storybook/cli sb init`
4. `nx g @nrwl/angular:storybook-configuration ng-nest` which fails with an error "Could not find module "@nrwl/storybook" from "/Users/leifwells/Code/nx-test-workspace".
5. `npm i -D @nrwl/storybook` to avoid the error
6. `nx g @nrwl/angular:storybook-configuration ng-nest` which ends with the error "Could not read apps/ng-nest/tsconfig.lib.json"

### Context

Please provide any relevant information about your setup:
```
  @nrwl/angular : 8.11.2
  @nrwl/cli : 8.11.2
  @nrwl/cypress : 8.11.2
  @nrwl/eslint-plugin-nx : Not Found
  @nrwl/express : Not Found
  @nrwl/jest : 8.11.2
  @nrwl/linter : 8.11.2
  @nrwl/nest : 8.11.2
  @nrwl/next : Not Found
  @nrwl/node : 8.11.2
  @nrwl/react : Not Found
  @nrwl/schematics : Not Found
  @nrwl/tao : 8.11.2
  @nrwl/web : Not Found
  @nrwl/workspace : 8.11.2
  typescript : 3.5.3
```

### PR DESCRIPTION
**Pull request by webberwang**
## Current Behavior
`@nrwl/storybook:configuration` only picks up `tsconfig.lib.json` & doesn't work for applications.

Described in #2382 & #3264 

## Expected Behavior

Pick up `tsconfig.app.json`

## Related Issue(s)

Fixes #2382, #3264

## Additional Problems

1) In the unit test, I used `@nrwl/react:application`, which generates `tsconfig.app.json`. But issue will arise when using `@nrwl/next:application`, which generates `tsconfig.json` instead of `tsconfig.app.json`

2) The `generateStories` only pick up from `lib` directory. In the unit test I used `@nrwl/react:component` to generate a component, which defaults to `lib`. If you have components under `app`, that won't be picked up.

### CODE REVIEW

**Comment by mandarini** [comment]
Hi @webberwang !

Thank you very much for this PR!

As you mentioned, `generateStories` only generates `.stories.*` files in `/lib` directories, and not in `/app` directories! You could fix this issue by checking out the `stories` schematic for `angular` and `react` for example. If you take a look at line 49 in [this file](https://github.com/nrwl/nx/blob/master/packages/react/src/schematics/stories/stories.ts), for example, you will notice that it chooses the `lib` directory. In this place, you can check the `projectType`, just like you did in the `configuration.ts` file you changed, and make it choose the `app` or the `lib` folder, accordingly! If this is fixed in all the projects' `stories` schematics, I think we may have a good result to move forward!

Would you like to try this out?

One more suggestion:
If you take a look at `schema.json` in the `storybook-configuration` schematics, where the prompts for the console are written, you can see that the `"generateStories"` prompt says `Automatically generate story files for components declared in this library?`. If we are to change the schematic to accommodate applications as well, maybe that text could be adjusted, too!

Let me know what you think, and also let me know if you need any help or guidance! I'm happy to help!

Thanks again! 😃

**Comment by webberwang** [comment]
@mandarini Let me know if everything looks good

**Review by juristr — CHANGES_REQUESTED** [review]
Tested on a react workspace: when I generate a new react workspace and then run `nx g @nrwl/react:storybook-configuration` on the `demoapp` (or some other arbitrary react app)

✅ it properly handles the fact that Cypress is already set up
✅ it properly generates Cypress specs
✅ it properly generates stories
✅ it properly generates story Cypress specs
❌ doesn't generate a `.storybook` folder in the `demoapp`.

That last issue then causes a run failure if I exec `nx run demoapp:storybook`

image

I also tried to do the same for an Angular workspace, but it doesn't seem to be handled there 🙂

**Comment by mandarini** [comment]
@juristr I cannot seem to fix the failing e2e test, any feedback/tips would be useful!

File: package.json [hunk 1]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "@nrwl/nx-source",
-  "version": "10.1.0",
+  "version": "999.1.1",
```
Comment (webberwang) [inline package.json:3]: is this meant to be committed remotely?

File: packages/react/src/schematics/storybook-configuration/schema.json [hunk 2]
```diff
@@ -5,7 +5,7 @@
   "properties": {
     "name": {
       "type": "string",
-      "description": "Library name",
+      "description": "Library or application name",
```
[the remark this answers is not in the document]
Comment (webberwang) [reply packages/react/src/schematics/storybook-configuration/schema.json:8]: I see I'll take a look

File: package.json (hunk 1 above)
Comment (mandarini) [reply package.json]: No, thank you, I will fix it!!

**Comment by webberwang** [comment]
> @juristr I cannot seem to fix the failing e2e test, any feedback/tips would be useful!
> The whole Storybook e2e suite fails with:
> 
> ```
> ENOENT: no such file or directory, stat 'libs/test-ui-lib-react7659083/.storybook'
> ```
> 
> which is strange because the lib `.storybook` directory seems to be generating correctly, and it also passes the unit tests..

Seems that https://github.com/nrwl/nx/pull/3611/commits/0f1ebc4116150e81e97f53c3589dfea606106ee9 introduced the e2e test errors

**Comment by webberwang** [comment]
> @juristr I cannot seem to fix the failing e2e test, any feedback/tips would be useful!
> The whole Storybook e2e suite fails with:
> 
> ```
> ENOENT: no such file or directory, stat 'libs/test-ui-lib-react7659083/.storybook'
> ```
> 
> which is strange because the lib `.storybook` directory seems to be generating correctly, and it also passes the unit tests..

To test locally, do I just run `yarn e2e`? It seems to get stuck

**Review by webberwang — COMMENTED** [review]
@mandarini I ended up replacing the edits you made to fix the e2e errors on circleci (force-pushed)

File: packages/storybook/src/schematics/configuration/configuration.ts [hunk 3]
```diff
@@ -93,12 +93,20 @@ function createLibStorybookDir(
...
+  const { projectType } = getProjectConfig(tree, projectName);
+  const projectPath = getProjectConfig(tree, projectName).root;
+  return join(
+    projectPath,
+    projectType === 'application' ? 'tsconfig.app.json' : 'tsconfig.lib.json'
+  );
+}
+
 function configureTsLibConfig(schema: StorybookConfigureSchema): Rule {
   const { name: projectName } = schema;
```
Comment (webberwang) [inline packages/storybook/src/schematics/configuration/configuration.ts:107]: This fixes Storybook config not being generated in React application

**Comment by mandarini** [comment]
@webberwang thanks, that's great! Reviewing now! :)

**Comment by mandarini** [comment]
> @mandarini I ended up replacing the edits you made to fix the e2e errors on circleci (force-pushed)

So, the e2e errors were fixed, but Nx broke. It has the same issues that it had before I made my changes.
1. Does not find `lib` folder in app
2. Cannot render `svg` due to missing webpack configuration
3. Fails when generating app Storybook, due to existing e2e tests

I will bring back the changes I made to fix these errors, and we should find a way to fix the failing e2e tests! If you can help in any way, that would be great. I will push my reverted fixes in a bit!

**Comment by webberwang** [comment]
> > @mandarini I ended up replacing the edits you made to fix the e2e errors on circleci (force-pushed)
> 
> Hi again @webberwang !
> 
> So, the e2e errors were fixed, but Nx broke. It has the same issues that it had before I made my changes.
> 
> 1. Does not find `lib` folder in app
> 2. Cannot render `svg` due to missing webpack configuration
> 3. Fails when generating app Storybook, due to existing e2e tests
> 
> I will bring back the changes I made to fix these errors, and we should find a way to fix the failing e2e tests! If you can help in any way, that would be great. I will push my reverted fixes in a bit!

I think our issue is some missing or incorrect tests. So we'll need to identify where that is

**Comment by mandarini** [comment]
Hi @webberwang !

So, for some reason I cannot find the commits I had made to address the issues I mentioned above, and I'm pretty confused! The commits were made on September 1st. 

In any case, here are the commands I am running:

### Create workspace with React
1. `npx create-nx-workspace@YOUR.LOCAL.VERSION worskpace-name`
2. `cd workspace-name`
3. `yarn add --dev <path_to_your_local_nx>/build/packages/storybook`
4. `yarn nx g @nrwl/react:lib ui-test`
5. `yarn nx g @nrwl/react:component test-lib-component --project=ui-test --export`
6. `yarn nx g @nrwl/react:component test-app-component --project=test-app`
7. `yarn nx g @nrwl/react:storybook-configuration --name=test-app` - error if I say yes to e2e tests, success if I say no to e2e tests
8. `yarn nx g @nrwl/react:storybook-configuration --name=ui-test` success
9. `nx run ui-test:storybook` success
10. `nx run test-app:storybook` error 1 - ` Can't resolve '../src/lib'` (see your Nx lib-files config.js, it's always `lib`, does not check project type). After I manually change config.js to reference `src/app` error: cannot render svg because it's not added in webpack

The thing is, I am pretty positive I had fixed and pushed these in a commit on September 1st - which indeed broke the tests.

Angular issues in next comment

**Comment by webberwang** [comment]
@mandarini I wasn't sure if you saved the branch locally but I made a backup before force pushing. https://github.com/webberwang/nx/tree/storybook-application-type-backup

Let me give those steps a try

**Comment by mandarini** [comment]
@webberwang oooh!!! Phew!! Ok, sorry, I panicked there for a moment! I've changed three machines in the past month, so "locally" does not exist for me, lol! Thanks! So, this commit https://github.com/webberwang/nx/commit/0f1ebc4116150e81e97f53c3589dfea606106ee9 fixed all the issues I mentioned above!

**Comment by webberwang** [comment]
> @webberwang oooh!!! Phew!! Ok, sorry, I panicked there for a moment! I've changed three machines in the past month, so "locally" does not exist for me, lol! Thanks! So, this commit [webberwang@0f1ebc4](https://github.com/webberwang/nx/commit/0f1ebc4116150e81e97f53c3589dfea606106ee9) fixed all the issues I mentioned above!

Great thanks I'll try to cherry pick those commits in there, will work on it tomorrow!

**Comment by mandarini** [comment]
Yes, we should definitely update the e2e tests... For some reason they are breaking, some .storybook folders are not getting generated... Let me know if you need any help there, I'll try my best!

**Comment by webberwang** [comment]
> Hi @webberwang !
> 
> So, for some reason I cannot find the commits I had made to address the issues I mentioned above, and I'm pretty confused! The commits were made on September 1st.
> 
> In any case, here are the commands I am running:
> 
> ### Create workspace with React
> 1. `npx create-nx-workspace@YOUR.LOCAL.VERSION worskpace-name`
> 2. `cd workspace-name`
> 3. `yarn add --dev <path_to_your_local_nx>/build/packages/storybook`
> 4. `yarn nx g @nrwl/react:lib ui-test`
> 5. `yarn nx g @nrwl/react:component test-lib-component --project=ui-test --export`
> 6. `yarn nx g @nrwl/react:component test-app-component --project=test-app`
> 7. `yarn nx g @nrwl/react:storybook-configuration --name=test-app` - error if I say yes to e2e tests, success if I say no to e2e tests
> 8. `yarn nx g @nrwl/react:storybook-configuration --name=ui-test` success
> 9. `nx run ui-test:storybook` success
> 10. `nx run test-app:storybook` error 1 - ` Can't resolve '../src/lib'` (see your Nx lib-files config.js, it's always `lib`, does not check project type). After I manually change config.js to reference `src/app` error 2 : cannot render svg because it's not added in webpack
> 
> The thing is, I am pretty positive I had fixed and pushed these in a commit on September 1st - which indeed broke the tests.
> 
> Angular issues in next comment

I'm guessing missing this?
4.5. yarn nx g @nrwl/react:application ui-test

**Comment by webberwang** [comment]
@mandarini Could I have your slack username in nrwl?

**Comment by webberwang** [comment]
Hey @mandarini,

This https://github.com/nrwl/nx/pull/3611/commits/f784bc0f650e1cd50d7f217ab4a3ca05760deb8e seems to fix the issue. Renaming the template directory for `./lib-files` to `./project-files` without updating the assets glob pattern was stopping `.storybook` from generating.

It likely worked on the local machine from previously generated artifacts.

**Comment by mandarini** [comment]
Hi @webberwang ! How is it going? Do you need any help?

**Comment by webberwang** [comment]
@mandarini It should be fixed in the last commit, let me know if I missed something thanks!

**Comment by mandarini** [comment]
@juristr I think this looks good, works as expected and we could merge it!

**Comment by ruslanguns** [comment]
Release date?

**Pull request merged**